database/optionally.go

```diff
@@ -0,0 +1,130 @@
+package database
+
+import (
+	"context"
+	"fmt"
+	"github.com/icinga/icinga-go-library/com"
+	"github.com/pkg/errors"
+)
+
+// Upsert upserts entities to the database.
+type Upsert interface {
+	// Stream bulk upserts the specified entities via NamedBulkExec.
+	// The upsert statement is created using BuildUpsertStmt with the
+	// first entity from the entities stream.
+	Stream(ctx context.Context, entities <-chan Entity) error
+}
+
+type ExecOption func(u *upsert)
+
+// WithOnUpsert adds callback for successful (bulk) DML operations.
+func WithOnUpsert(onUpsert ...OnSuccess[Entity]) ExecOption {
+	return func(u *upsert) {
+		u.onUpsert = onUpsert
+	}
+}
+
+// WithStatement uses own instead of the standard upsert statement
+func WithStatement(stmt string, placeholders int) ExecOption {
```

Comment:
WithStatement uses the specified statement for bulk upserts instead of automatically creating one.